yt_dlp/extractor/niconico.py

```diff
@@ -367,229 +388,120 @@ def parse_format_id(id_code):
     def _real_extract(self, url):
         video_id = self._match_id(url)
 
-        # Get video webpage for API data.
-        webpage, handle = self._download_webpage_handle(
-            'http://www.nicovideo.jp/watch/' + video_id, video_id)
-        if video_id.startswith('so'):
-            video_id = self._match_id(handle.geturl())
-
-        api_data = self._parse_json(self._html_search_regex(
-            'data-api-data="([^"]+)"', webpage,
-            'API data', default='{}'), video_id)
-
-        def get_video_info_web(items):
-            return dict_get(api_data['video'], items)
-
-        # Get video info
-        video_info_xml = self._download_xml(
-            'http://ext.nicovideo.jp/api/getthumbinfo/' + video_id,
-            video_id, note='Downloading video info page')
-
-        def get_video_info_xml(items):
-            if not isinstance(items, list):
-                items = [items]
-            for item in items:
-                ret = xpath_text(video_info_xml, './/' + item)
-                if ret:
-                    return ret
-
-        if get_video_info_xml('error'):
-            error_code = get_video_info_xml('code')
-
-            if error_code == 'DELETED':
-                raise ExtractorError('The video has been deleted.',
-                                     expected=True)
-            elif error_code == 'NOT_FOUND':
-                raise ExtractorError('The video is not found.',
-                                     expected=True)
-            elif error_code == 'COMMUNITY':
-                self.to_screen('%s: The video is community members only.' % video_id)
-            else:
-                raise ExtractorError('%s reports error: %s' % (self.IE_NAME, error_code))
+        try:
+            webpage, handle = self._download_webpage_handle(
+                'http://www.nicovideo.jp/watch/' + video_id, video_id)
+            if video_id.startswith('so'):
+                video_id = self._match_id(handle.geturl())
+
+            api_data = self._parse_json(self._html_search_regex(
+                'data-api-data="([^"]+)"', webpage,
+                'API data', default='{}'), video_id)
+        except ExtractorError as e:
+            try:
+                api_data = self._download_json(
+                    'https://www.nicovideo.jp/api/watch/v3/%s?_frontendId=6&_frontendVersion=0&actionTrackId=AAAAAAAAAA_%d' % (video_id, time_millis()), video_id,
+                    note='Downloading API JSON', errnote='Unable to fetch data')['data']
+            except (ExtractorError, KeyError):
+                if not isinstance(e.cause, compat_HTTPError):
+                    raise e
+                else:
+                    e = e.cause
+                webpage = e.read().decode('utf-8', 'replace')
```

Comment:
```suggestion
                e = e.cause
                webpage = e.read().decode('utf-8', 'replace')
```
or even just
```suggestion
                webpage = e.cause.read().decode('utf-8', 'replace')
```
based on whether u really need `raise e` below